roachtest: version/mixed/nodes=5 failed

SHA: https://github.com/cockroachdb/cockroach/commits/77f26d185efb436aaac88243de19a27caa5da9b6

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=version/mixed/nodes=5 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1509340&tab=artifacts#/version/mixed/nodes=5

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190926-1509340/version/mixed/nodes=5/run_1
	cluster.go:2143,version.go:233,version.go:246,test_runner.go:689: unexpected node event: 2: dead

```

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x1bd9e44]

goroutine 176 [running]:
panic(0x3b493a0, 0x6f39c70)
	/usr/local/go/src/runtime/panic.go:565 +0x2c5 fp=0xc003608c90 sp=0xc003608c00 pc=0x78d165
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000025560, 0x494fca0, 0xc0007181e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x121 fp=0xc003608cf0 sp=0xc003608c90 pc=0x12d0211
runtime.call32(0x0, 0x417d478, 0xc00049a490, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:519 +0x3b fp=0xc003608d20 sp=0xc003608cf0 pc=0x7bb1db
panic(0x3b493a0, 0x6f39c70)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5 fp=0xc003608db0 sp=0xc003608d20 pc=0x78d055
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:82
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:390 +0x411 fp=0xc003608de0 sp=0xc003608db0 pc=0x7a2cc1
github.com/cockroachdb/cockroach/pkg/storage.(*replicaAppBatch).runPreApplyTriggers(0xc0009370c8, 0x494fca0, 0xc003a93f50, 0xc00c6f2008, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_application_state_machine.go:679 +0x374 fp=0xc003609050 sp=0xc003608de0 pc=0x1bd9e44
github.com/cockroachdb/cockroach/pkg/storage.(*replicaAppBatch).Stage(0xc0009370c8, 0x492de20, 0xc00c6f2008, 0xc0036093c8, 0x1bd526b, 0xc0009372a8, 0xc0009372d8)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_application_state_machine.go:468 +0x2fd fp=0xc003609390 sp=0xc003609050 pc=0x1bd8f8d
github.com/cockroachdb/cockroach/pkg/storage/apply.Batch.Stage-fm(0x492de20, 0xc00c6f2008, 0xc00c6f2008, 0xc0009372a8, 0x498a500, 0xc0009372a8)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/apply/task.go:69 +0x43 fp=0xc0036093d8 sp=0xc003609390 pc=0x1b3f803
github.com/cockroachdb/cockroach/pkg/storage/apply.mapCmdIter(0x498a500, 0xc0009372a8, 0xc0036094c8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/apply/cmd.go:162 +0x11b fp=0xc003609438 sp=0xc0036093d8 pc=0x1b3e73b
github.com/cockroachdb/cockroach/pkg/storage/apply.(*Task).applyOneBatch(0xc003609908, 0x494fca0, 0xc003a93f50, 0x498a500, 0xc000937278, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/apply/task.go:270 +0x15d fp=0xc0036094f0 sp=0xc003609438 pc=0x1b3f1fd
github.com/cockroachdb/cockroach/pkg/storage/apply.(*Task).ApplyCommittedEntries(0xc003609908, 0x494fca0, 0xc003a93f50, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/apply/task.go:247 +0xcf fp=0xc003609548 sp=0xc0036094f0 pc=0x1b3f01f
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc000937000, 0x494fca0, 0xc003a93f50, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raft.go:760 +0xdbd fp=0xc003609c78 sp=0xc003609548 pc=0x1c06d6d
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue.func1(0x494fca0, 0xc003a93f50, 0xc000937000, 0x494fca0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3761 +0x12e fp=0xc003609d38 sp=0xc003609c78 pc=0x1c7bf7e
github.com/cockroachdb/cockroach/pkg/storage.(*Store).withReplicaForRequest(0xc000b50700, 0x494fca0, 0xc003a93f50, 0xc003bcc800, 0xc003609ec0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3514 +0x16b fp=0xc003609dc0 sp=0xc003609d38 pc=0x1c4422b
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc000b50700, 0x494fca0, 0xc0007181e0, 0x48)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3750 +0x1f4 fp=0xc003609ef8 sp=0xc003609dc0 pc=0x1c44f44
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc001092700, 0x494fca0, 0xc0007181e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:238 +0x214 fp=0xc003609f50 sp=0xc003609ef8 pc=0x1c31264
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x494fca0, 0xc0007181e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x3e fp=0xc003609f78 sp=0xc003609f50 pc=0x1c7665e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc000436f70, 0xc000025560, 0xc000436f60)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0xfb fp=0xc003609fc8 sp=0xc003609f78 pc=0x12d23ab
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc003609fd0 sp=0xc003609fc8 pc=0x7bcef1
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xa8
```

This seems new.

--- a/pkg/storage/replica_application_state_machine.go
+++ b/pkg/storage/replica_application_state_machine.go
@@ -668,6 +668,16 @@ func (b *replicaAppBatch) runPreApplyTriggers(ctx context.Context, cmd *replicat
 		b.r.mu.Unlock()
 		b.changeRemovesReplica = true
 
+		// In 19.1 and before we used DeprecatedNextReplicaID to carry the next
+		// replica id to use once this change is applied. In 19.2 we started
+		// providing a new range descriptor directly, which includes this info.
+		var nextReplID roachpb.ReplicaID
+		if change.Desc != nil {
+			nextReplID = change.Desc.NextReplicaID
+		} else {
+			nextReplID = change.DeprecatedNextReplicaID
+		}
+
 		// Delete all of the local data. We're going to delete the hard state too.
 		// In order for this to be safe we need code above this to promise that we're
 		// never going to write hard state in response to a message for a later
@@ -676,7 +686,7 @@ func (b *replicaAppBatch) runPreApplyTriggers(ctx context.Context, cmd *replicat
 			ctx,
 			b.batch,
 			b.batch,
-			change.Desc.NextReplicaID,
+			nextReplID,
 			false, /* clearRangeIDLocalOnly */
 			false, /* mustUseClearRange */
 		); err != nil {

--- a/pkg/storage/replica_application_state_machine_test.go
+++ b/pkg/storage/replica_application_state_machine_test.go
@@ -1,0 +1,152 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage/storagepb"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/stretchr/testify/require"
+	"go.etcd.io/etcd/raft/raftpb"
+)
+
+// TestReplicaStateMachineStageChangeReplicas tests the behavior of staging a
+// replicated command with a ChangeReplicas trigger in a replicaAppBatch. The
+// test exercises the logic of applying both old-style and new-style
+// ChangeReplicas triggers.
+func TestReplicaStateMachineStageChangeReplicas(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	tc := testContext{}
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.TODO())
+	tc.Start(t, stopper)
+
+	// Lock the replica for the entire test.
+	r := tc.repl
+	r.raftMu.Lock()
+	defer r.raftMu.Unlock()
+	sm := r.getStateMachine()
+
+	desc := r.Desc()
+	replDesc, ok := desc.GetReplicaDescriptor(r.store.StoreID())
+	require.True(t, ok)
+
+	makeCmd := func(desc *roachpb.RangeDescriptor, t roachpb.ChangeReplicasTrigger) *replicatedCmd {
+		return &replicatedCmd{
+			ctx: context.Background(),
+			ent: &raftpb.Entry{Index: r.mu.state.RaftAppliedIndex + 1},
+			decodedRaftEntry: decodedRaftEntry{
+				idKey: makeIDKey(),
+				raftCmd: storagepb.RaftCommand{
+					ProposerLeaseSequence: r.mu.state.Lease.Sequence,
+					MaxLeaseIndex:         r.mu.state.LeaseAppliedIndex + 1,
+					ReplicatedEvalResult: storagepb.ReplicatedEvalResult{
+						State:          &storagepb.ReplicaState{Desc: desc},
+						ChangeReplicas: &storagepb.ChangeReplicas{ChangeReplicasTrigger: t},
+						Timestamp:      r.mu.state.GCThreshold.Add(1, 0),
+					},
+				},
+			},
+		}
+	}
+
+	t.Run("add replica", func(t *testing.T) {
+		// Add a new replica to the Range.
+		newDesc := *desc
+		newDesc.InternalReplicas = append([]roachpb.ReplicaDescriptor(nil), desc.InternalReplicas...)
+		addedReplDesc := newDesc.AddReplica(replDesc.NodeID+1, replDesc.StoreID+1, roachpb.VOTER_FULL)
+
+		testutils.RunTrueAndFalse(t, "deprecated", func(t *testing.T, deprecated bool) {
+			var trigger roachpb.ChangeReplicasTrigger
+			if deprecated {
+				trigger = roachpb.ChangeReplicasTrigger{
+					DeprecatedChangeType: roachpb.ADD_REPLICA,
+					DeprecatedReplica:    addedReplDesc,
+					DeprecatedUpdatedReplicas: []roachpb.ReplicaDescriptor{
+						replDesc,
+						addedReplDesc,
+					},
+					DeprecatedNextReplicaID: addedReplDesc.ReplicaID + 1,
+				}
+			} else {
+				trigger = roachpb.ChangeReplicasTrigger{
+					Desc:                  &newDesc,
+					InternalAddedReplicas: []roachpb.ReplicaDescriptor{addedReplDesc},
+				}
+			}
+
+			// Create a new application batch.
+			b := sm.NewBatch(false /* ephemeral */).(*replicaAppBatch)
+			defer b.Close()
+
+			// Stage a command with the ChangeReplicas trigger.
+			cmd := makeCmd(&newDesc, trigger)
+			_, err := b.Stage(cmd)
+			require.NoError(t, err)
+			require.False(t, b.changeRemovesReplica)
+			require.Equal(t, b.state.RaftAppliedIndex, cmd.ent.Index)
+			require.Equal(t, b.state.LeaseAppliedIndex, cmd.raftCmd.MaxLeaseIndex)
+
+			// Check the replica's destroy status.
+			r.mu.Lock()
+			require.False(t, r.mu.destroyStatus.Removed())
+			r.mu.Unlock()
+		})
+	})
+
+	t.Run("remove replica", func(t *testing.T) {
+		// Remove ourselves from the Range.
+		newDesc := *desc
+		newDesc.InternalReplicas = append([]roachpb.ReplicaDescriptor(nil), desc.InternalReplicas...)
+		removedReplDesc, ok := newDesc.RemoveReplica(replDesc.NodeID, replDesc.StoreID)
+		require.True(t, ok)
+
+		testutils.RunTrueAndFalse(t, "deprecated", func(t *testing.T, deprecated bool) {
+			var trigger roachpb.ChangeReplicasTrigger
+			if deprecated {
+				trigger = roachpb.ChangeReplicasTrigger{
+					DeprecatedChangeType:      roachpb.REMOVE_REPLICA,
+					DeprecatedReplica:         removedReplDesc,
+					DeprecatedUpdatedReplicas: []roachpb.ReplicaDescriptor{},
+					DeprecatedNextReplicaID:   replDesc.ReplicaID + 1,
+				}
+			} else {
+				trigger = roachpb.ChangeReplicasTrigger{
+					Desc:                    &newDesc,
+					InternalRemovedReplicas: []roachpb.ReplicaDescriptor{removedReplDesc},
+				}
+			}
+
+			// Create a new application batch.
+			b := sm.NewBatch(false /* ephemeral */).(*replicaAppBatch)
+			defer b.Close()
+
+			// Stage a command with the ChangeReplicas trigger.
+			cmd := makeCmd(&newDesc, trigger)
+			_, err := b.Stage(cmd)
+			require.NoError(t, err)
+			require.True(t, b.changeRemovesReplica)
+			require.Equal(t, b.state.RaftAppliedIndex, cmd.ent.Index)
+			require.Equal(t, b.state.LeaseAppliedIndex, cmd.raftCmd.MaxLeaseIndex)
+
+			// Check the replica's destroy status.
+			r.mu.Lock()
+			require.True(t, r.mu.destroyStatus.Removed())
+			r.mu.destroyStatus.Set(nil, destroyReasonAlive) // reset
+			r.mu.Unlock()
+		})
+	})
+}

